### DISCUSSION TITLE
VisionOs support for existing React Native app

### DISCUSSION THREAD

**Discussion by mpatafio** [callstack/react-native-visionos#108] [Q&A]
Hi all!

is it possible to integrate react-native-visionos within an existing React Native app?

Thanks in advance.

**Comment by okwasniewski** [comment]
Hey! 

Yes, it's possible. Currently, there is no automated way to do this (but we have this on our backlog).

Here are the steps you need to take: 

1. Generate a new project and copy `visionos` folder to an existing one. 
2. Add `@callstack/react-native-visionos` and `@callstack/out-of-tree-platforms` as dependencies
3. Modify `metro.config.js` and add custom resolver:

```js
const {getDefaultConfig, mergeConfig} = require('@react-native/metro-config');
const {getPlatformResolver} = require('@callstack/out-of-tree-platforms');

const config = {
  resolver: {
    resolveRequest: getPlatformResolver({
      platformNameMap: {visionos: '@callstack/react-native-visionos'},
    }),
  },
};

module.exports = mergeConfig(getDefaultConfig(__dirname), config);
```

4. Add `exclude` to `tsconfig.json`

```json
"exclude": [
    "**/Pods/**",
]
```

You might run into some issues with your dependencies not supporting visionOS, In that case you can try quickly patching their podspecs using method described in this blog post: https://www.oskarkwasniewski.dev/blog/bringing-react-native-libraries-to-apple-vision-pro

**Reply by MP185496** [reply]
Is it required to use the latest version of RN in our project in order for this to work?

**Reply by okwasniewski** [reply]
@MP185496 Preferably yes, but it may work if you are using old architecture. There might be some issues when using Codegen

**Reply by MP185496** [reply]
@okwasniewski mhm thanks, so right now in our project we are using RN: 0.71.4 version. so i can just copy visionos folder to the app, follow the steps you wrote here and hopefully it will work? Or to do this minimal RN version is 0.73? Thanks

**Reply by okwasniewski** [reply]
Yes, you can try following these steps and let me know if that works

**Reply by MP185496** [reply]
Thanks, i did that and now I have problem with bundle exec pod install
[!] Invalid `Podfile` file: Unsupported platform `visionos`. Platform must be `:ios`, `:osx`, `:macos`, `:tvos`, or `:watchos`..

 #  from /Users/user/Documents/5.3/project/apps/mobile/visionos/Podfile:8
 #  -------------------------------------------
 #  
 >  platform :visionos, min_visionos_version_supported
 #  prepare_react_native_project!
 #  -------------------------------------------

**Reply by okwasniewski** [reply]
Hey @MP185496, can you provide minimal reproduction for this? I have plans to create a full guide in our docs (that will be created soon)

**Reply by MP185496** [reply]
Sure:
In my existing project I installed @callstack/react-native-visionos and @callstack/out-of-tree-platforms 
Made new project with npx @callstack/react-native-visionos@latest init Mobile
Copied visionos folder from that project into my project

Added  in metro.config.js
resolveRequest: getPlatformResolver({
      platformNameMap: {visionos: '@callstack/react-native-visionos'},
    }),
    
Added in tsconfig.json
"exclude": [
    "**/Pods/**",
]
Searched in node_modules for platform support but couldn't find any s.platforms in them
changed folder path of node_modules in Podefile of visionos folder to be on location where are my node_modules

**Reply by milanp93** [reply]
@okwasniewski Hey any thoughts on this, did you have time to make the guide for this use case? Thanks

**Reply by okwasniewski** [reply]
@milanp93 Yes, there you go: https://callstack.github.io/react-native-visionos-docs/docs/guides/integrate-into-existing-projects

**Reply by milanp93** [reply]
@okwasniewski Thanks! is there any workaround when we use some of the expo packages?

**Reply by okwasniewski** [reply]
Unfortunately, not at the moment
